File: android/app/src/main/java/com/mar/addapp/UsageTrackerPlugin.java
package com.mar.addapp;

import android.app.AppOpsManager;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Process;
import android.provider.Settings;

import com.getcapacitor.JSArray;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import org.json.JSONArray;

@CapacitorPlugin(name = "UsageTracker")
public class UsageTrackerPlugin extends Plugin {

    public static final String PREFS_NAME        = "UsageTrackerPrefs";
    public static final String KEY_TRACKED_APPS  = "tracked_apps";
    public static final String KEY_PENDING_SHAME = "pending_shame";
    public static final String KEY_NOTIFIED_TODAY = "notified_today";

    private SharedPreferences prefs() {
        return getContext().getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
    }

    @PluginMethod
    public void checkUsagePermission(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("granted", hasUsagePermission());
        call.resolve(ret);
    }

    @PluginMethod
    public void requestUsagePermission(PluginCall call) {
        Intent intent = new Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        getContext().startActivity(intent);
        call.resolve();
    }

    @PluginMethod
    public void startTracking(PluginCall call) {
        JSArray appsArray = call.getArray("apps");
        if (appsArray == null) {
            call.reject("apps parameter required");
            return;
        }
        prefs().edit()
            .putString(KEY_TRACKED_APPS, appsArray.toString())
            .putString(KEY_NOTIFIED_TODAY, "[]")
            .apply();

        Intent intent = new Intent(getContext(), UsageTrackerService.class);
        getContext().startForegroundService(intent);

        JSObject ret = new JSObject();
        ret.put("started", true);
        call.resolve(ret);
    }

    @PluginMethod
    public void stopTracking(PluginCall call) {
        Intent intent = new Intent(getContext(), UsageTrackerService.class);
        getContext().stopService(intent);
        call.resolve();
    }

    @PluginMethod
    public void getTodayUsageMinutes(PluginCall call) {
        String packageName = call.getString("packageName");
        if (packageName == null) {
            call.reject("packageName required");
            return;
        }
        if (!hasUsagePermission()) {
            JSObject ret = new JSObject();
            ret.put("minutes", -1);
            ret.put("permissionRequired", true);
            call.resolve(ret);
            return;
        }
        long minutes = UsageTrackerService.getTodayUsageMinutes(getContext(), packageName);
        JSObject ret = new JSObject();
        ret.put("minutes", minutes);
        call.resolve(ret);
    }

    @PluginMethod
    public void getPendingShameData(PluginCall call) {
        String json = prefs().getString(KEY_PENDING_SHAME, "[]");
        try {
            JSArray arr = new JSArray(json);
            JSObject ret = new JSObject();
            ret.put("sessions", arr);
            call.resolve(ret);
        } catch (Exception e) {
            JSObject ret = new JSObject();
            ret.put("sessions", new JSArray());
            call.resolve(ret);
        }
    }

    @PluginMethod
    public void clearPendingShameData(PluginCall call) {
        prefs().edit().putString(KEY_PENDING_SHAME, "[]").apply();
        call.resolve();
    }

    @PluginMethod
    public void isTracking(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("active", UsageTrackerService.isRunning);
        call.resolve(ret);
    }

    private boolean hasUsagePermission() {
        AppOpsManager appOps = (AppOpsManager) getContext()
            .getSystemService(Context.APP_OPS_SERVICE);
        if (appOps == null) return false;
        @SuppressWarnings("deprecation")
        int mode = appOps.checkOpNoThrow(
            AppOpsManager.OPSTR_GET_USAGE_STATS,
            Process.myUid(),
            getContext().getPackageName()
        );
        return mode == AppOpsManager.MODE_ALLOWED;
    }
}

File: android/app/src/main/java/com/mar/addapp/UsageTrackerPlugin.kt
package com.mar.addapp

import android.app.AppOpsManager
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.provider.Settings
import app.tauri.annotation.Command
import app.tauri.annotation.InvokeArg
import com.getcapacitor.JSArray
import com.getcapacitor.JSObject
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
import com.getcapacitor.annotation.CapacitorPlugin
import org.json.JSONArray
import org.json.JSONObject

@CapacitorPlugin(name = "UsageTracker")
class UsageTrackerPlugin : Plugin() {

    companion object {
        const val PREFS_NAME       = "UsageTrackerPrefs"
        const val KEY_TRACKED_APPS = "tracked_apps"
        const val KEY_PENDING_SHAME = "pending_shame"
        const val KEY_NOTIFIED_TODAY = "notified_today"
    }

    private fun prefs(): SharedPreferences =
        context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)

    // ── Check if Usage Access permission is granted ───────────
    @PluginMethod
    fun checkUsagePermission(call: PluginCall) {
        val granted = hasUsagePermission()
        val ret = JSObject()
        ret.put("granted", granted)
        call.resolve(ret)
    }

    // ── Open the Usage Access settings screen ─────────────────
    @PluginMethod
    fun requestUsagePermission(call: PluginCall) {
        val intent = Intent(Settings.ACTION_USAGE_ACCESS_SETTINGS)
        intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK
        activity.startActivity(intent)
        call.resolve()
    }

    // ── Start tracking ────────────────────────────────────────
    // Expects: { apps: [ { packageName, appName, limitMinutes, majorGoal, emoji } ] }
    @PluginMethod
    fun startTracking(call: PluginCall) {
        val appsArray = call.getArray("apps") ?: run {
            call.reject("apps parameter required")
            return
        }

        // Persist tracked app list for the service to read
        prefs().edit().putString(KEY_TRACKED_APPS, appsArray.toString()).apply()

        // Reset notified-today set so limits can fire again
        prefs().edit().putString(KEY_NOTIFIED_TODAY, "[]").apply()

        // Start the foreground service
        val intent = Intent(context, UsageTrackerService::class.java)
        context.startForegroundService(intent)

        val ret = JSObject()
        ret.put("started", true)
        call.resolve(ret)
    }

    // ── Stop tracking ─────────────────────────────────────────
    @PluginMethod
    fun stopTracking(call: PluginCall) {
        val intent = Intent(context, UsageTrackerService::class.java)
        context.stopService(intent)
        call.resolve()
    }

    // ── Get today's usage for a specific package ──────────────
    // Expects: { packageName: "com.zhiliaoapp.musically" }
    @PluginMethod
    fun getTodayUsageMinutes(call: PluginCall) {
        val packageName = call.getString("packageName") ?: run {
            call.reject("packageName required")
            return
        }
        if (!hasUsagePermission()) {
            val ret = JSObject()
            ret.put("minutes", -1)
            ret.put("permissionRequired", true)
            call.resolve(ret)
            return
        }
        val minutes = UsageTrackerService.getTodayUsageMinutes(context, packageName)
        val ret = JSObject()
        ret.put("minutes", minutes)
        call.resolve(ret)
    }

    // ── Get all pending shame events ──────────────────────────
    @PluginMethod
    fun getPendingShameData(call: PluginCall) {
        val json = prefs().getString(KEY_PENDING_SHAME, "[]") ?: "[]"
        val arr  = JSArray(json)
        val ret  = JSObject()
        ret.put("sessions", arr)
        call.resolve(ret)
    }

    // ── Clear pending shame events ────────────────────────────
    @PluginMethod
    fun clearPendingShameData(call: PluginCall) {
        prefs().edit().putString(KEY_PENDING_SHAME, "[]").apply()
        call.resolve()
    }

    // ── Is service currently running ──────────────────────────
    @PluginMethod
    fun isTracking(call: PluginCall) {
        val ret = JSObject()
        ret.put("active", UsageTrackerService.isRunning)
        call.resolve(ret)
    }

    // ── Helpers ───────────────────────────────────────────────
    private fun hasUsagePermission(): Boolean {
        val appOps = context.getSystemService(Context.APP_OPS_SERVICE) as AppOpsManager
        val mode = appOps.unsafeCheckOpNoThrow(
            AppOpsManager.OPSTR_GET_USAGE_STATS,
            android.os.Process.myUid(),
            context.packageName
        )
        return mode == AppOpsManager.MODE_ALLOWED
    }
}
